components/common/src/main/java/com/hotels/styx/config/schema/Schema.java

```diff
@@ -232,23 +370,74 @@ public FieldValue elementType() {
+        @Override
+        public Optional<String> isCorrectType(List<String> parents, JsonNode parent, JsonNode element, Function<String, FieldValue> schemas) {
+            return element.isObject() ? validateObject(parents, parent, element, schemas) : message(parents, element);
+        }
+
+        @Override
+        public String describe() {
+            return format("OBJECT (%s)", schema.name());
+        }
+
+        private Optional<String> message(List<String> parents, JsonNode tree) {
+            String fullName = Joiner.on(".").join(parents);
+            return Optional.of(format("Unexpected field type. Field '%s' should be OBJECT ('%s'), but it is %s", fullName, this.schema.name(), tree.getNodeType()));
+        }
+
+        private Optional<String> validateObject(List<String> parents, JsonNode parent, JsonNode tree, Function<String, FieldValue> lookup) {
+
+            if (schema.ignore()) {
+                return Optional.empty();
+            }
+
+            for (Schema.Field field : schema.fields()) {
+                if (isMandatory(schema, field) && tree.get(field.name()) == null) {
+                    throw new SchemaValidationException(format("Missing a mandatory field '%s.%s'", Joiner.on(".").join(parents), field.name()));
+                }
 
-        ObjectFieldLazy(String schemaName) {
-            this.schemaName = requireNonNull(schemaName);
+                if (tree.get(field.name()) != null) {
+                    String name = field.name();
+                    Optional<String> correctType = field.value().isCorrectType(push(parents, name), tree, tree.get(name), lookup);
+                    if (correctType.isPresent()) {
+                        throw new SchemaValidationException(correctType.get());
+                    }
+                }
+            }
+
+            schema.constraints().forEach(constraint -> {
+                if (!constraint.evaluate(schema, tree)) {
+                    throw new SchemaValidationException("Schema constraint failed. " + constraint.describe());
+                }
+            });
+
+            assertNoUnknownFields(Joiner.on(".").join(parents), schema, ImmutableList.copyOf(tree.fieldNames()));
+
+            return Optional.empty();
         }
 
-        public String schemaName() {
-            return schemaName;
+        private static boolean isMandatory(Schema schema, Schema.Field field) {
+            return !schema.optionalFields().contains(field.name());
         }
     }
 
+    private static void assertNoUnknownFields(String prefix, Schema schema, List<String> fieldsPresent) {
+        Set<String> knownFields = ImmutableSet.copyOf(schema.fieldNames());
+
+        fieldsPresent.forEach(name -> {
+            if (!knownFields.contains(name)) {
+                throw new SchemaValidationException(format("Unexpected field: '%s.%s'", prefix, name));
+            }
+        });
+    }
+
+    static List<String> push(List<String> currentList, String element) {
```

Comment:
```suggestion
    private static List<String> concatenate(List<String> currentList, String element) {
```

Reply:
+1 for private.
But no for `concatenate`. This is supposed to behave like a stack, and therefore `push` is more appropriate.